cpu/nrf52/periph/pwm.c

```diff
@@ -0,0 +1,184 @@
+#define CNTR_MODE_MASK  (0x0001)
+#define MAX_PRESCALER   (7U)
+
+/** @} */
+
+/**
+ * @brief   When calculating the actual frequency, we allow for a deviation of
+ *          ~15.5% (val +- (val / 8))
+ */
+#define F_DEV           (8)
+
+/**
+ * @brief   Allocate some memory for the PWM sequences
+ */
+static uint16_t pwm_seq[PWM_NUMOF][PWM_CHANNELS];
+
+static inline NRF_PWM_Type *dev(pwm_t pwm)
+{
+    return pwm_config[pwm].dev;
+}
+
+uint32_t pwm_init(pwm_t pwm, pwm_mode_t mode, uint32_t freq, uint16_t res)
+{
+    uint32_t real_clk;
+
+    /* check if given device is valid */
+    if ((pwm >= PWM_NUMOF) || (res > MAX_RES)) {
+        return 0;
+    }
+
+    /* make sure the device is stopped */
+    dev(pwm)->TASKS_STOP = 1;
+    dev(pwm)->ENABLE = 0;
+
+    /* calculate the needed frequency, for center modes we need double */
+    uint32_t clk = ((freq * res) << (mode & PWM_CENTER));
+    /* match to best fitting prescaler */
+    for (uint8_t i = 0; i < (MAX_PRESCALER + 1); i++) {
+        real_clk = (PERIPH_CLOCK >> i);
+        if (((real_clk - (real_clk / F_DEV)) < clk) &&
+            ((real_clk + (real_clk / F_DEV)) > clk)) {
+            dev(pwm)->PRESCALER = i;
+            break;
+        }
+        if (i == MAX_PRESCALER) {
+            return 0;
+        }
+    }
+    real_clk /= res;
+
+    /* pin configuration */
+    for (unsigned i = 0; i < PWM_CHANNELS; i++) {
+        if (pwm_config[pwm].pin[i] != GPIO_UNDEF) {
+            NRF_P0->PIN_CNF[pwm_config[pwm].pin[i]] = PIN_CNF_SET;
+        }
+        /* either left aligned pol or inverted duty cycle */
+        pwm_seq[pwm][i] = (POL_MASK & mode) ? POL_MASK : res;
+        dev(pwm)->PSEL.OUT[i] = pwm_config[pwm].pin[i];
+        DEBUG("set PIN[%i] to %i with 0x%x\n",
+              (int)i, (int)pwm_config[pwm].pin[i], pwm_seq[pwm][i]);
+    }
+
+    /* enable the device */
+    dev(pwm)->ENABLE = 1;
+
+    /* finally get the actual selected frequency */
+    DEBUG("set real f to %i\n", (int)real_clk);
+    dev(pwm)->COUNTERTOP = (res - 1);
+
+    /* select PWM mode */
+    dev(pwm)->MODE = (mode & CNTR_MODE_MASK);
+    dev(pwm)->LOOP = 0;
+    dev(pwm)->DECODER = PWM_DECODER_LOAD_Individual;
+
+    DEBUG("MODE: 0x%08x\n", (int)dev(pwm)->MODE);
+    DEBUG("LOOP: 0x%08x\n", (int)dev(pwm)->LOOP);
+    DEBUG("DECODER: 0x%08x\n", (int)dev(pwm)->DECODER);
+
+    /* setup the sequence */
+    dev(pwm)->SEQ[0].PTR = (uint32_t)pwm_seq[pwm];
+    dev(pwm)->SEQ[0].CNT = pwm_channels(pwm);
+    dev(pwm)->SEQ[0].REFRESH = 0;
+    dev(pwm)->SEQ[0].ENDDELAY = 0;
+
+    DEBUG("ptr: 0x%08x\n", (int)dev(pwm)->SEQ[0].PTR);
+    DEBUG("cnt: 0x%08x\n", (int)dev(pwm)->SEQ[0].CNT);
+    DEBUG("refresh: 0x%08x\n", (int)dev(pwm)->SEQ[0].REFRESH);
+    DEBUG("enddelay: 0x%08x\n", (int)dev(pwm)->SEQ[0].ENDDELAY);
+
+    /* start sequence */
+    dev(pwm)->TASKS_SEQSTART[0] = 1;
+
+    DEBUG("PWM started\n");
+
+    return real_clk;
```

Comment:
I assume this value has to be adapted in case of  `PWM_CENTER` ?
```
init 0 1 10000 100
The pwm frequency is set to 10000

init 0 2 10000 100
The pwm frequency is set to 20000
```

However, frequency is 10k in both cases (as expected)

Reply:
@PeterKietzmann i intend on dropping center mode support. I already checked that there is no support in center mode for other CPUs, also center mode is rearly used anyway. If center mode has no special relevance to you, i will just drop it.

Reply:
I'm okay with dropping that feature.